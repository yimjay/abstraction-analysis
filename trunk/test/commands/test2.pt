include ExpStmt.incl
include Loops.incl
<* code templates *>
<code foo pars=(abc) >
  FOO abc
</code>

<eval DEBUG{
    3 : FLOAT;
    "3.3" : FLOAT;
}/>

<input syntax="Cfront.code" to=result >
  int b; //this is comment
  /* what is this */
  int main() {}
</input>

<input syntax="Cfront.code" to=result parse=_ >
    -3 //@ =>EXP
    -3 //@ =>EXP|""
    3.2e5 //@ =>EXP
    3.2a5 //@ => (EXP,_)
    3.2 //@ =>EXP
    .1 //@ =>EXP
    int abc[3]; //@=>Stmt
    static char listname_base[MAXLINE]; //@=>Stmt
    ABC abc; //@=>Stmt
    ABC (*p)(); //@ =>Stmt
    a+b-c //@=>EXP;
    beta * C[j*ldc+i] //@ => EXP;
    for(i=0;i<100;i+=2) //@ => var=Loop; 
    -1+N //@ => EXP
    -*N //@ => EXP
    3.7 //@ => EXP
    mat[0][i] != 0 //@ => EXP
</input>
<eval PRINT("parsing result=" result); />

<eval DEBUG{
    if (("GOO" "abc") ==> cur1=foo) foo#(x) = cur1;
    if (("FOO" "abc") ==> cur2=foo) foo#(x) = cur2;
    a = Stmt#"abdc";
    Stmt#a = a; 
    a = ("a" "b" "c");
    b = (a "d");
    REPLACE(a,b,a);
    SPLIT("","for(i=0;i<100;i+=2)");
    SPLIT("","(abc def)") => LIST(STRING,""); 
    PERMUTE((3 2 1), ("a" "b" "c"));
    SPLIT("","a+b-c");
    REPLACE("x","y", SPLIT("","x*x-2")); 
    REPLACE( (("a",1) ("b",2) ("c",3)), SPLIT("","a+b-c"));
    1 * ArrayAccess#("A","i"*70+"l");
}/>

<eval DEBUG{
    input = Stmt#"var";
    DUPLICATE("var", (1 2 3), input);

    input : Stmt#_;
    Stmt#input : Stmt#_;
    Stmt#input : Stmt#newvar;
    Stmt#123 : Stmt#newvar;
}/>

