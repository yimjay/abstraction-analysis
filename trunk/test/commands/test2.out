
DEBUGGING at level 1
3:FLOAT
===>0
"3.3":FLOAT
===>1
END DEBUGGING at level 1
parsing result=Uop#(
  "-",
  3)
"
"Uop#(
  "-",
  3)
"
"EXPO#(
  "3.2",
  "5")
"
"(
"3.2",
"a5")
"
""3.2""
"".1""
"DeclStmt#TypeInfo#(
    ArrayType#(
      IntType#"int",
      ArrayDim#3),
    "abc",
    "")
"
"DeclStmt#TypeInfo#(
    ArrayType#(
      StaticType#IntType#"char",
      ArrayDim#"MAXLINE"),
    "listname_base",
    "")
"
"DeclStmt#TypeInfo#(
    "ABC",
    "abc",
    "")
"
"FunctionDecl#(
  "p",
  "",
  "ABC",
  EmptyStmt)#(
  "",
  "",
  "")
"
"Bop#(
  "-",
  Bop#(
    "+",
    "a",
    "b"),
  "c")
"
"Bop#(
  "*",
  "beta",
  ArrayAccess#(
    "C",
    Bop#(
      "+",
      Bop#(
        "*",
        "j",
        "ldc"),
      "i")))
"
"var={Loop#(
  "i",
  0,
  "100",
  2)#(
  "",
  "<",
  "",
  "",
  "",
  "",
  "",
  "<")}
"
"Bop#(
  "+",
  Uop#(
    "-",
    1),
  "N")
"
"Uop#(
  "-",
  Uop#(
    "*",
    "N"))
"
""3.7""
"Bop#(
  "!=",
  ArrayAccess#(
    ArrayAccess#(
      "mat",
      0),
    "i"),
  0)
"
" NULL

DEBUGGING at level 1
("GOO" "abc")==>cur1=foo
===>0
if ("GOO" "abc")==>cur1=foo===>0
("FOO" "abc")==>cur2=foo
===>1
if ("FOO" "abc")==>cur2=foo===>1
foo#x=cur2
===>foo#abc
a=Stmt#"abdc"
===>Stmt#"abdc"
Stmt#a=a
===>Stmt#"abdc"
a=("a" "b" "c")
===>("a" "b" "c")
b=(a "d")
===>(("a" "b" "c") "d")
REPLACE(a,b,a)
===>(("a" "b" "c") "d")
SPLIT("","for(i=0;i<100;i+=2)")
===>("for" "(" "i" "=" "0" ";" "i" "<" "100" ";" "i" "+" "=" "2" ")")
SPLIT("","(abc def)")
===>("(" "abc" " " "def" ")")
SPLIT("","(abc def)")=>LIST(STRING,"")
===>("(" "abc" "def" ")")
PERMUTE((3 2 1),("a" "b" "c"))
===>("c" "b" "a")
SPLIT("","a+b-c")
===>("a" "+" "b" "-" "c")
SPLIT("","x*x-2")
===>("x" "*" "x" "-" "2")
REPLACE("x","y",SPLIT("","x*x-2"))
===>("y" "*" "y" "-" "2")
SPLIT("","a+b-c")
===>("a" "+" "b" "-" "c")
REPLACE((("a",1) ("b",2) ("c",3)),SPLIT("","a+b-c"))
===>(1 "+" 2 "-" 3)
Result of invoking RebuildBop(op,opd1,opd2)
===>CodeVar:Bop#("*",70,"i")
Result of invoking RebuildBop(op,opd1,opd2)
===>CodeVar:Bop#("+",Bop#("*",70,"i"),"l")
Result of invoking RebuildBop(op,opd1,opd2)
===>CodeVar:ArrayAccess#("A",Bop#("+",Bop#("*",70,"i"),"l"))
END DEBUGGING at level 1

DEBUGGING at level 1
input=Stmt#"var"
===>Stmt#"var"
DUPLICATE("var",(1 2 3),input)
===>(Stmt#1 Stmt#2 Stmt#3)
input:Stmt#_
===>1
Stmt#input:Stmt#_
===>1
Stmt#input:Stmt#newvar
===>1
Stmt#123:Stmt#newvar
===>0
END DEBUGGING at level 1
